File: src/xkeygrab.ts
// int type, Window window, Window subwindow, int x, int y, Time time, unsigned int keycode, unsigned int modifiers

import { FFIType, JSCallback, Pointer } from "bun:ffi";
import { WindowID, WindowRef } from "./WindowRef";
import { ffi } from "./lib";
import { Ref } from "./ref";

const Mod2Mask = 1 << 1;

interface KeyGrabEvent {
  // type: number;
  key: number;
  modifiers: number;
  window: WindowRef;
  subwindow: WindowRef;
  pos: { x: number; y: number };
  /** milliseconds relative to start of X11 */
  time: number;
}

const callback = new JSCallback(
  (
    type: number,
    window: WindowID,
    subwindow: WindowID,
    x: number,
    y: number,
    time: number,
    key: number,
    modifiers: number
  ) => {
    const hash = (key << 8) + (modifiers & ~Mod2Mask);
    const m = map.get(window);
    if (m) {
      const xkeygrab = m.get(hash);
      if (xkeygrab) {
        xkeygrab.callback({
          key,
          modifiers,
          get window() {
            return new WindowRef(window);
          },
          get subwindow() {
            return new WindowRef(subwindow);
          },
          pos: { x, y },
          time,
        });
      }
    }
  },
  {
    args: [
      FFIType.int,
      FFIType.u64_fast,
      FFIType.u64_fast,
      FFIType.int,
      FFIType.int,
      FFIType.u32,
      FFIType.u64_fast,
      FFIType.u64_fast,
    ],
    threadsafe: true,
  }
);

ffi.jskeygrab_set_cb(callback.ptr!);

const map = new Map<WindowID, Map<number, XKeyGrab>>();

class XKeyGrab {
  #hash: number;
  #ptr: Pointer;
  #ref: Ref;
  readonly key: number;
  readonly modifiers: number;
  readonly window: WindowID = 0;
  readonly callback: (event: any) => void;

  constructor(
    options: { key: number; modifiers: number; window?: WindowID | WindowRef },
    callback: (event: any) => void
  ) {
    let { key, modifiers, window = 0 } = options;
    if (typeof window === "object") {
      window = window.id;
    }
    const hash = (key << 8) + (modifiers & ~Mod2Mask);
    const m = map.get(window);
    if (!m) {
      map.set(window, new Map([[hash, this]]));
    } else {
      if (m.has(hash)) {
        throw new Error(
          "XKeygrab already exists for this key and modifier combination."
        );
      }
    }

    this.#ptr = ffi.jskeygrab_add(key, modifiers, window);
    if (!this.#ptr) {
      throw new Error("Failed to create XKeyGrab");
    }

    this.#hash = hash;
    this.key = key;
    this.modifiers = modifiers;
    this.window = window;
    this.callback = callback;
    this.#ref = new Ref();
  }

  dispose() {
    const m = map.get(this.window);
    if (m) {
      m.delete(this.#hash);
      if (m.size === 0) {
        map.delete(this.window);
      }
    }
    ffi.jskeygrab_dispose(this.#ptr);
    this.#ref.unref();
  }
}

const keygrab = new XKeyGrab(
  { key: 67 /* f1 */, modifiers: 1 << 6 /* mod4 (super) */ },
  (event) => {
    console.log("keygrab event", event);
  }
);

File: src/lib/XKeyGrab.zig
const std = @import("std");
const global = @import("./global.zig");
const Window = X.Window;
const X = global.X;
const allocator = global.allocator;

const KeyGrab = struct {
    keycode: i32,
    modifiers: u32,
    window: Window,
    hasMod2: bool,
};
const KeyGrabCallback = fn (type: i32, window: Window, subwindow: Window, x: i32, y: i32, time: X.Time, keycode: u32, modifiers: u32) void;

var count: u32 = 0;
var callback: ?*KeyGrabCallback = null;
var thread: ?std.Thread = null;

fn keygrabThread() void {
    var event: X.XEvent = undefined;
    while (count > 0) {
        _ = X.XNextEvent(global.dpy, &event);
        if (event.type == X.KeyPress) {
            callback.?(event.xkey.type, if (event.xkey.window == event.xkey.root) 0 else event.xkey.window, event.xkey.subwindow, event.xkey.x, event.xkey.y, event.xkey.time, event.xkey.keycode, event.xkey.state);
        }
    }
}

pub fn setCallback(cb: *KeyGrabCallback) callconv(.C) void {
    callback = cb;
}

pub fn add(keycode: i32, modifiers: u32, window: Window) callconv(.C) ?*KeyGrab {
    var w = if (window == 0) X.DefaultRootWindow(global.dpy) else window;

    var result = X.XGrabKey(global.dpy, keycode, modifiers, w, 0, X.GrabModeAsync, X.GrabModeAsync);
    if (result != 1) {
        std.log.err("XGrabKey returned {d}", .{result});
        return null;
    }

    const key = allocator.create(KeyGrab) catch {
        std.log.err("failed to allocate KeyGrab", .{});
        return null;
    };
    key.keycode = keycode;
    key.modifiers = modifiers;
    key.window = w;

    if ((modifiers & X.Mod2Mask) == 0) {
        result = X.XGrabKey(global.dpy, keycode, modifiers | X.Mod2Mask, w, 0, X.GrabModeAsync, X.GrabModeAsync);
        key.hasMod2 = result == 1;
    } else {
        key.hasMod2 = false;
    }

    if (count == 0) {
        thread = std.Thread.spawn(.{}, keygrabThread, .{}) catch {
            std.log.err("failed to spawn thread", .{});
            allocator.destroy(key);
            return null;
        };
    }
    count += 1;

    return key;
}

pub fn dispose(key: *KeyGrab) callconv(.C) void {
    _ = X.XUngrabKey(global.dpy, key.keycode, key.modifiers, key.window);
    if (key.hasMod2) _ = X.XUngrabKey(global.dpy, key.keycode, key.modifiers | X.Mod2Mask, key.window);
    allocator.destroy(key);
    count -= 1;
    if (count == 0) {
        thread.?.join();
        thread = null;
    }
}
